File: tests/fixtures/python/vulnerable.py
import os
import pickle
import random
import subprocess
import yaml

# SQL Injection - f-string in execute
def login(username, password):
    query = f"SELECT * FROM users WHERE username='{username}' AND password='{password}'"
    db.execute(query)

# SQL Injection - %-formatting
def get_user(user_id):
    cursor.execute("SELECT * FROM users WHERE id=%s" % user_id)

# OS Command Injection
def ping_host(host):
    os.system("ping -c 1 " + host)

# OS Command Injection - subprocess with shell=True
def run_command(cmd):
    subprocess.call(cmd, shell=True)

# Hardcoded secrets
API_KEY = "sk-1234567890abcdef1234567890abcdef"
password = "SuperSecret123!"

# Insecure random
def generate_token():
    return random.randint(100000, 999999)

# Eval/Exec
def evaluate(expr):
    result = eval(expr)
    return result

def run_code(code):
    exec(code)

# Path traversal
def download_file(filename):
    path = "/uploads/" + filename
    return open(path + ".txt")

# Insecure deserialization
def load_data(data):
    obj = pickle.loads(data)
    return obj

# Insecure YAML loading
def load_config(config_str):
    config = yaml.load(config_str)
    return config

# XSS in template
from markupsafe import Markup
def render_name(name):
    return Markup("<b>" + name + "</b>")

# Log injection
import logging
def log_user_action(request):
    logging.info("User action: " + request.path)

File: tests/fixtures/javascript/vulnerable.js
const http = require('http');
const fs = require('fs');
const path = require('path');

// XSS - innerHTML
function renderComment(comment) {
    document.getElementById('comments').innerHTML = comment;
}

// XSS - document.write
function showMessage(msg) {
    document.write(msg);
}

// Eval injection
function calculate(expression) {
    return eval(expression);
}

// Eval via Function constructor
function runCode(code) {
    const fn = new Function(code);
    return fn();
}

// setTimeout with string (eval-like)
function delayed(code) {
    setTimeout("alert('hello')", 1000);
}

// Prototype pollution
function merge(target, source) {
    for (const key in source) {
        target[key] = source[key];
    }
}

// Hardcoded secrets
const API_KEY = "sk-1234567890abcdef1234567890abcdef";
const password = "SuperSecretPassword123";

// Insecure random
function generateId() {
    return Math.random().toString(36);
}

// Path traversal
function readUserFile(req, res) {
    const filePath = path.join('/uploads', req.query.filename);
    fs.readFile(filePath, (err, data) => {
        res.send(data);
    });
}

// SSRF
function proxyRequest(req, res) {
    fetch(req.query.url).then(response => {
        res.send(response);
    });
}

// NoSQL injection
async function findUser(req, res) {
    const user = await db.collection('users').findOne({ username: req.body.username });
    res.json(user);
}

// Insecure deserialization
const serialize = require('node-serialize');
function processData(req, res) {
    const obj = serialize.unserialize(req.body.data);
    res.json(obj);
}

// dangerouslySetInnerHTML (React)
function Comment({ text }) {
    return <div dangerouslySetInnerHTML={{ __html: text }} />;
}
